daceml/onnx/onnx_importer.py

```diff
@@ -348,60 +419,84 @@ def __call__(
+            if input in self.sdfg.free_symbols:
                 clean_inputs[input] = arr
             else:
                 clean_inputs[clean_onnx_name(input)] = arr
 
         # add the weights
         params = {}
         for name, arr in self.weights.items():
-            if len(arr.shape) == 0:
-                params[clean_onnx_name(name)] = arr[()]
+            desc = self.sdfg.arrays[clean_onnx_name(name)]
+            if type(desc) is dt.Scalar:
+                params[clean_onnx_name(name)] = arr.cpu().numpy()[()]
             else:
-                params[clean_onnx_name(name)] = arr.copy()
+                params[clean_onnx_name(name)] = arr.clone()
 
-        inferred_symbols = infer_symbols_from_shapes(sdfg, {
+        inferred_symbols = infer_symbols_from_shapes(self.sdfg, {
             **clean_inputs,
             **params
         })
-        # TODO @orausch if this is removed the SDFG complains
-        # TypeError: Type mismatch for argument ONNX_unk__493: expected scalar type, got <class 'sympy.core.numbers.Integer'>
-        # fix this better
         inferred_symbols = {k: int(v) for k, v in inferred_symbols.items()}
 
-        def eval_dim(dim):
-            for sym in dim.free_symbols:
-                dim = dim.subs(sym, inferred_symbols[sym.name])
-            return dim
+        if torch_outputs is None:
+            torch_outputs = any(
+                isinstance(inp, torch.Tensor)
+                for _, inp in clean_inputs.items())
 
         outputs = OrderedDict()
         # create numpy arrays for the outputs
         for output in self.outputs:
             clean_name = clean_onnx_name(output)
-            arr = sdfg.arrays[clean_name]
-
-            # TODO @orausch add error handling for evalf
-            shape = [
-                eval_dim(d) if type(d) is dace.symbol else d for d in arr.shape
-            ]
-            outputs[clean_name] = np.empty(shape,
-                                           dtype=arr.dtype.as_numpy_dtype())
-
-        sdfg.expand_library_nodes()
-
-        if self.apply_strict:
-            sdfg.apply_strict_transformations()
-
-        sdfg(**clean_inputs, **params, **outputs, **inferred_symbols)
-
-        if len(outputs) == 1:
-            return next(iter(outputs.values()))
-
-        return tuple(outputs.values())
+            outputs[clean_name] = create_output_array(
+                inferred_symbols,
+                self.sdfg.arrays[clean_name],
+                use_torch=torch_outputs)
+
+        # check that there's no overlap
+        seen = set()
+        for parameters in [clean_inputs, params, outputs, inferred_symbols]:
+            new_parameters = set(parameters)
+            assert not seen.intersection(new_parameters)
+            seen |= new_parameters
+
+        return clean_inputs, params, inferred_symbols, outputs
+
+
+def create_output_array(
+        inferred_symbols: typing.Dict[str, int],
+        desc: dt.Data,
+        use_torch=False,
```

Comment:
shouldn't this be =True by default? I may be wrong, please explain why.

Reply:
The default value is never used anyway. I thought `False` is better since dace is so numpy oriented anyway. For example, I also made it so that `ONNXModel`s return numpy arrays if only numpy inputs are passed.